multipaz-models/src/commonMain/kotlin/org/multipaz/models/provisioning/ProvisioningModel.kt

```diff
@@ -0,0 +1,366 @@
+import org.multipaz.provision.openid4vci.OpenID4VCI
+import org.multipaz.provision.openid4vci.OpenID4VCIBackend
+import org.multipaz.provision.openid4vci.OpenID4VCIClientPreferences
+import org.multipaz.rpc.backend.BackendEnvironment
+import org.multipaz.sdjwt.credential.KeyBoundSdJwtVcCredential
+import org.multipaz.sdjwt.credential.KeylessSdJwtVcCredential
+import org.multipaz.securearea.CreateKeySettings
+import org.multipaz.securearea.SecureArea
+import org.multipaz.securearea.SecureAreaProvider
+import org.multipaz.util.Logger
+import org.multipaz.util.toBase64Url
+import kotlin.coroutines.CoroutineContext
+import kotlin.math.min
+import kotlin.reflect.KClass
+import kotlin.reflect.safeCast
+import kotlin.time.Clock
+import kotlin.time.Duration.Companion.days
+
+/**
+ * This model supports UX/UI flow for provisioning of credentials.
+ *
+ * Only a single provisioning session per model object can be active at any time.
+ *
+ * @param documentStore new [Document] will be created in this [DocumentStore]
+ * @param secureArea [Credential] objects will be bound to this [SecureArea]
+ * @param httpClient HTTP client used to communicate to the provisioning server, it MUST NOT
+ *     handle redirects automatically
+ * @param promptModel [PromptModel] that is used to show prompts to generate proof-of-possession for
+ *     credential keys
+ * @param documentMetadataInitializer function that initializes document metadata at [Document]
+ *     creation time based on issuance server metadata
+ */
+class ProvisioningModel(
+    private val documentStore: DocumentStore,
+    private val secureArea: SecureArea,
+    private val httpClient: HttpClient,
+    private val promptModel: PromptModel,
+    private val documentMetadataInitializer: suspend (
+            metadata: AbstractDocumentMetadata,
+            credentialDisplay: Display,
+            issuerDisplay: Display
+        ) -> Unit
+) {
+    private var mutableState = MutableStateFlow<State>(Idle)
+
+    /** State of the model */
+    val state: StateFlow<State> get() = mutableState.asStateFlow()
+
+    private val authorizationResponseChannel = Channel<AuthorizationResponse>()
+
+    private var job: Job? = null
+
+    /**
+     * Launch provisioning session to provision credentials to a new [Document] using
+     * OpenID4VCI protocol.
+     *
+     * @param offerUri credential offer (formatted as URI with custom protocol name)
+     * @param clientPreferences configuration parameters for OpenID4VCI client
+     * @param backend interface to the wallet back-end service
+     * @return deferred [Document] value
+     */
+    fun launchOpenID4VCIProvisioning(
+        offerUri: String,
+        clientPreferences: OpenID4VCIClientPreferences,
+        backend: OpenID4VCIBackend,
+    ): Deferred<Document> =
+        launch(Dispatchers.Default + promptModel + ProvisioningEnvironment(backend)) {
+            OpenID4VCI.createClientFromOffer(offerUri, clientPreferences)
+        }
```

Comment:
Is there a good reason to do `fun foo(): Deferred<Bar>` instead of `suspend fun foo(): Bar` we're doing everywhere else? Also, I think in general it's a code-smell to use a specific dispatcher in a library (`Dispatchers.Default` here), that needs to be either a parameter or this could be a suspend fun.